core/shared/src/main/scala/clue/ApolloClient.scala

```diff
@@ -35,29 +35,38 @@ protected[clue] trait Emitter[F[_]] {
 }
 
 // Client internal state for the FSM.
-protected sealed abstract class State[+F[_], +CP](val status: PersistentClientStatus)
+// We keep a connectionId throughout all states to ensure that callback events (onClose, onMessage)
+// correpond to the current connection iteration. This is important in case of reconnections.
+protected sealed abstract class State[+F[_], +CP](val status: PersistentClientStatus) {
+  val connectionId: ConnectionId
+}
 protected object State {
-  final case object Disconnected
+  final case class Disconnected(connectionId: ConnectionId)
       extends State[Nothing, Nothing](PersistentClientStatus.Disconnected)
-  final case class Connecting[F[_]](latch: Latch[F])
+  final case class Connecting[F[_]](connectionId: ConnectionId, latch: Latch[F])
       extends State[F, Nothing](PersistentClientStatus.Connecting)
-  final case class Connected[F[_], CP](connection: PersistentConnection[F, CP])
-      extends State[F, CP](PersistentClientStatus.Connected)
+  final case class Connected[F[_], CP](
+    connectionId: ConnectionId,
+    connection:   PersistentConnection[F, CP]
+  ) extends State[F, CP](PersistentClientStatus.Connected)
   final case class Initializing[F[_], CP](
-    initPayloadF:  F[Map[String, Json]],
```

Comment:
Don't we send auth info in the initial payload?

Reply:
Yes, but we store it (or rather an effect that returns what the payload should be) in case the connections drops and we need to rebuild it.